sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -87,20 +87,25 @@ RT::PiProgram ProgramManager::getBuiltOpenCLProgram(OSModuleHandle M,
   std::shared_ptr<context_impl> Ctx = getSyclObjImpl(Context);
   std::map<OSModuleHandle, RT::PiProgram> &CachedPrograms =
       Ctx->getCachedPrograms();
-  auto It = CachedPrograms.find(M);
-  if (It != CachedPrograms.end())
-    return It->second;
 
-  DeviceImage *Img = nullptr;
-  using PiProgramT = remove_pointer_t<RT::PiProgram>;
-  unique_ptr_class<PiProgramT, decltype(RT::piProgramRelease)> ProgramManaged(
-      loadProgram(M, Context, &Img), RT::piProgramRelease);
+  {
+    std::lock_guard<std::mutex> CacheLock(Ctx->getProgramKernelCacheMutex());
 
-  build(ProgramManaged.get(), Img->BuildOptions);
-  RT::PiProgram Program = ProgramManaged.release();
-  CachedPrograms[M] = Program;
+    auto It = CachedPrograms.find(M);
+    if (It != CachedPrograms.end())
+      return It->second;
 
-  return Program;
+    DeviceImage *Img = nullptr;
+    using PiProgramT = remove_pointer_t<RT::PiProgram>;
+    unique_ptr_class<PiProgramT, decltype(RT::piProgramRelease)> ProgramManaged(
+        loadProgram(M, Context, &Img), RT::piProgramRelease);
+
+    build(ProgramManaged.get(), Img->BuildOptions);
+    RT::PiProgram Program = ProgramManaged.release();
```

Comment:
Can we exclude this code from the critical section?

User might submit different requests to build programs from several threads and this code effectively synchronizes the build at SYCL RT level, which doesn't seem necessary

Reply:
Done.
The same scenario could be applied here too: [#discussion_r350034908]